File: bindings/rust/yunlink/src/runtime.rs
use std::ffi::CString;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Mutex, MutexGuard};
use std::thread::{self, JoinHandle};
use std::time::Duration;

use tokio::sync::broadcast;
use yunlink_sys as sys;

use crate::configuration::{
    register_callbacks, ConfigurationCallbackContext, ConfigurationResponse,
};
use crate::error::{ensure, Result};
use crate::events::{parse_event, Event, EVENT_CHANNEL_CAPACITY};
use crate::ffi_util::write_c_buffer;
use crate::types::{
    AgentType, AuthorityLease, AuthorityState, CommandHandle, ControlSource, PeerConnection,
    RuntimeConfig, Session, SessionInfo, SessionState, TargetSelector,
};

mod commands;
mod configuration;

/// Raw C ABI runtime pointer.
///
#[derive(Debug, Clone, Copy)]
struct RawRuntime(*mut sys::yunlink_runtime_t);

// The C++ runtime is internally synchronized for the operations exposed by this
// wrapper. We still keep the pointer behind a Mutex in `Runtime` so all safe
// method calls retrieve it through one consistent path.
unsafe impl Send for RawRuntime {}
unsafe impl Sync for RawRuntime {}

impl RawRuntime {
    fn ptr(self) -> *mut sys::yunlink_runtime_t {
        self.0
    }
}

/// Safe Rust runtime facade over the bindings-oriented C ABI.
///
/// The object owns exactly one opaque `yunlink_runtime_t*`. Public methods use
/// Rust domain types, convert them into `yunlink_*_t` C structs, call the raw
/// `yunlink-sys` symbols, and map result codes back into `YunlinkError`.
pub struct Runtime {
    /// Opaque C ABI runtime handle.
    raw: Mutex<RawRuntime>,
    /// Signals the event polling thread to stop.
    shutdown: std::sync::Arc<AtomicBool>,
    /// Broadcast channel used by Rust subscribers.
    sender: broadcast::Sender<Event>,
    configuration_sender: broadcast::Sender<ConfigurationResponse>,
    configuration_callback_context: Box<ConfigurationCallbackContext>,
    configuration_tokens: Vec<usize>,
    /// Background thread that polls `yunlink_runtime_poll_event`.
    poll_thread: Option<JoinHandle<()>>,
}

impl Runtime {
    /// Create, configure, and start a YunLink runtime through C ABI.
    pub fn start(config: RuntimeConfig) -> Result<Self> {
        let mut raw_ptr = std::ptr::null_mut();
        ensure(unsafe { sys::yunlink_runtime_create(&mut raw_ptr) })?;

        // Build the C ABI config struct explicitly. The `Default`
        // implementation supplies `struct_size` and zeroed fixed string
        // buffers; the helper below then fills those buffers safely.
        let mut native_cfg = sys::yunlink_runtime_config_t {
            udp_bind_port: config.udp_bind_port,
            udp_target_port: config.udp_target_port,
            tcp_listen_port: config.tcp_listen_port,
            connect_timeout_ms: 5000,
            io_poll_interval_ms: 5,
            max_buffer_bytes_per_peer: 1 << 20,
            self_identity: sys::yunlink_identity_t {
                agent_type: config.agent_type.to_native(),
                agent_id: config.agent_id,
                role: match config.agent_type {
                    AgentType::GroundStation => sys::YUNLINK_ROLE_CONTROLLER,
                    AgentType::Uav | AgentType::Ugv => sys::YUNLINK_ROLE_VEHICLE,
                    AgentType::SwarmController => sys::YUNLINK_ROLE_CONTROLLER,
                    AgentType::Unknown(_) => sys::YUNLINK_ROLE_OBSERVER,
                },
            },
            security_tags_enabled: 1,
            ..Default::default()
        };
        write_c_buffer(&mut native_cfg.shared_secret, &config.shared_secret);
        write_c_buffer(&mut native_cfg.multicast_group, &config.multicast_group);

        ensure(unsafe { sys::yunlink_runtime_start(raw_ptr, &native_cfg) })?;

        let (sender, _) = broadcast::channel(EVENT_CHANNEL_CAPACITY);
        let (configuration_sender, _) = broadcast::channel(EVENT_CHANNEL_CAPACITY);
        let (configuration_callback_context, configuration_tokens) =
            match register_callbacks(raw_ptr, configuration_sender.clone()) {
                Ok(registration) => registration,
                Err(error) => {
                    let _ = unsafe { sys::yunlink_runtime_stop(raw_ptr) };
                    unsafe { sys::yunlink_runtime_destroy(raw_ptr) };
                    return Err(error);
                }
            };
        let shutdown = std::sync::Arc::new(AtomicBool::new(false));
        let thread_shutdown = shutdown.clone();
        let thread_sender = sender.clone();
        let thread_raw = RawRuntime(raw_ptr);

        let poll_thread = thread::spawn(move || {
            while !thread_shutdown.load(Ordering::Relaxed) {
                let mut event = sys::yunlink_runtime_event_t::default();
                let result =
                    unsafe { sys::yunlink_runtime_poll_event(thread_raw.ptr(), &mut event) };
                if result == sys::YUNLINK_RESULT_OK
                    && event.type_ != sys::YUNLINK_RUNTIME_EVENT_NONE
                {
                    if let Some(parsed) = parse_event(event) {
                        let _ = thread_sender.send(parsed);
                    }
                    continue;
                }
                thread::sleep(Duration::from_millis(10));
            }
        });

        Ok(Self {
            raw: Mutex::new(RawRuntime(raw_ptr)),
            shutdown,
            sender,
            configuration_sender,
            configuration_callback_context,
            configuration_tokens,
            poll_thread: Some(poll_thread),
        })
    }

    /// Subscribe to parsed runtime events.
    pub fn subscribe(&self) -> broadcast::Receiver<Event> {
        self.sender.subscribe()
    }

    /// Subscribe to owned configuration service responses.
    pub fn subscribe_configuration(&self) -> broadcast::Receiver<ConfigurationResponse> {
        self.configuration_sender.subscribe()
    }

    /// Connect to a remote peer through the C ABI TCP client helper.
    pub async fn connect(&self, ip: &str, port: u16) -> Result<PeerConnection> {
        let ip = CString::new(ip)?;
        let mut peer = sys::yunlink_peer_t::default();
        ensure(unsafe { sys::yunlink_peer_connect(self.raw_ptr(), ip.as_ptr(), port, &mut peer) })?;
        Ok(PeerConnection::from_raw(peer))
    }

    /// Open an active session with a connected peer.
    pub async fn open_session(&self, peer: &PeerConnection, node_name: &str) -> Result<Session> {
        let node_name = CString::new(node_name)?;
        let mut session = sys::yunlink_session_t::default();
        ensure(unsafe {
            sys::yunlink_session_open(self.raw_ptr(), &peer.raw, node_name.as_ptr(), &mut session)
        })?;
        Ok(Session {
            session_id: session.session_id,
        })
    }

    /// Describe an existing session, if the runtime still knows it.
    pub fn describe_session(&self, session: &Session) -> Result<Option<SessionInfo>> {
        let native_session = session.to_native();
        let mut info = sys::yunlink_session_info_t::default();
        let result =
            unsafe { sys::yunlink_session_describe(self.raw_ptr(), &native_session, &mut info) };
        if result == sys::YUNLINK_RESULT_NOT_FOUND {
            return Ok(None);
        }
        ensure(result)?;
        Ok(Some(SessionInfo {
            session_id: info.session_id,
            state: SessionState::from_native(info.state),
            remote_agent_type: AgentType::from_native(info.remote_identity.agent_type),
            remote_agent_id: info.remote_identity.agent_id,
            peer: PeerConnection::from_raw(info.peer),
            capability_flags: info.capability_flags,
            node_name: crate::ffi_util::string_from_c_buf(&info.node_name),
        }))
    }

    /// Request authority for a target.
    pub async fn request_authority(
        &self,
        peer: &PeerConnection,
        session: &Session,
        target: &TargetSelector,
        source: ControlSource,
        lease_ttl_ms: u32,
        allow_preempt: bool,
    ) -> Result<()> {
        let session = session.to_native();
        ensure(unsafe {
            sys::yunlink_authority_request(
                self.raw_ptr(),
                &peer.raw,
                &session,
                &target.raw,
                source.to_native(),
                lease_ttl_ms,
                if allow_preempt { 1 } else { 0 },
            )
        })
    }

    /// Release authority for a target.
    pub async fn release_authority(
        &self,
        peer: &PeerConnection,
        session: &Session,
        target: &TargetSelector,
    ) -> Result<()> {
        let session = session.to_native();
        ensure(unsafe {
            sys::yunlink_authority_release(self.raw_ptr(), &peer.raw, &session, &target.raw)
        })
    }

    /// Renew authority for a target.
    pub async fn renew_authority(
        &self,
        peer: &PeerConnection,
        session: &Session,
        target: &TargetSelector,
        source: ControlSource,
        lease_ttl_ms: u32,
    ) -> Result<()> {
        let session = session.to_native();
        ensure(unsafe {
            sys::yunlink_authority_renew(
                self.raw_ptr(),
                &peer.raw,
                &session,
                &target.raw,
                source.to_native(),
                lease_ttl_ms,
            )
        })
    }

    /// Return the current authority lease observed by this runtime.
    pub fn current_authority(&self) -> Result<Option<AuthorityLease>> {
        let mut lease = sys::yunlink_authority_lease_t::default();
        let result = unsafe { sys::yunlink_authority_current(self.raw_ptr(), &mut lease) };
        if result == sys::YUNLINK_RESULT_NOT_FOUND {
            return Ok(None);
        }
        ensure(result)?;
        Ok(Some(AuthorityLease {
            state: AuthorityState::from_native(lease.state),
            session_id: lease.session_id,
            peer: PeerConnection::from_raw(lease.peer),
        }))
    }

    /// Return the current opaque runtime pointer.
    pub(crate) fn raw_ptr(&self) -> *mut sys::yunlink_runtime_t {
        self.raw_lock().0
    }

    /// Lock the raw runtime handle.
    fn raw_lock(&self) -> MutexGuard<'_, RawRuntime> {
        self.raw.lock().expect("raw runtime mutex poisoned")
    }

    /// Convert a raw C ABI command handle into the public safe handle.
    fn command_handle_from_native(handle: sys::yunlink_command_handle_t) -> CommandHandle {
        CommandHandle {
            session_id: handle.session_id,
            message_id: handle.message_id,
            correlation_id: handle.correlation_id,
        }
    }
}

impl Drop for Runtime {
    fn drop(&mut self) {
        // Stop the polling thread before destroying the opaque runtime. This
        // avoids polling a pointer after `yunlink_runtime_destroy`.
        self.shutdown.store(true, Ordering::Relaxed);
        if let Some(handle) = self.poll_thread.take() {
            let _ = handle.join();
        }
        let raw = self.raw.lock().expect("raw runtime mutex poisoned").0;
        for token in self.configuration_tokens.drain(..) {
            let _ = unsafe { sys::yunlink_configuration_unsubscribe(raw, token) };
        }
        let _ = unsafe { sys::yunlink_runtime_stop(raw) };
        unsafe { sys::yunlink_runtime_destroy(raw) };
        let _ = &self.configuration_callback_context;
    }
}

File: bindings/python/src/yunlink/runtime.py
from __future__ import annotations

import asyncio
import queue
import threading
import time

from ._yunlink_native import RuntimeCore
from .errors import YunlinkError, call_native
from .events import ErrorEvent, coerce_event
from .configuration import coerce_configuration_response
from .configuration_runtime import ConfigurationRuntimeMixin
from .types import (
    AuthorityLease,
    CommandHandle,
    ControlSource,
    GotoCommand,
    PeerConnection,
    RuntimeConfig,
    Session,
    TargetSelector,
    VehicleCoreState,
)


class Runtime(ConfigurationRuntimeMixin):
    def __init__(self, core: RuntimeCore):
        self._core = core
        self._stop = threading.Event()
        self._poll_error: YunlinkError | None = None
        self._subscribers: list[queue.Queue] = []
        self._async_subscribers: list[tuple[asyncio.AbstractEventLoop, asyncio.Queue]] = []
        self._configuration_subscribers: list[queue.Queue] = []
        self._async_configuration_subscribers: list[
            tuple[asyncio.AbstractEventLoop, asyncio.Queue]
        ] = []
        self._thread = threading.Thread(target=self._poll_loop, daemon=True)
        self._thread.start()

    @classmethod
    def start(cls, config: RuntimeConfig) -> "Runtime":
        core = call_native(RuntimeCore)
        call_native(core.start, config.to_native())
        return cls(core)

    def close(self) -> None:
        if self._stop.is_set():
            return
        self._stop.set()
        self._thread.join(timeout=1.0)
        call_native(self._core.stop)

    def subscribe(self) -> queue.Queue:
        q: queue.Queue = queue.Queue()
        self._subscribers.append(q)
        return q

    def subscribe_async(self) -> asyncio.Queue:
        q: asyncio.Queue = asyncio.Queue()
        loop = asyncio.get_running_loop()
        self._async_subscribers.append((loop, q))
        return q

    def subscribe_configuration(self) -> queue.Queue:
        q: queue.Queue = queue.Queue()
        self._configuration_subscribers.append(q)
        return q

    def subscribe_configuration_async(self) -> asyncio.Queue:
        q: asyncio.Queue = asyncio.Queue()
        loop = asyncio.get_running_loop()
        self._async_configuration_subscribers.append((loop, q))
        return q

    def last_poll_error(self) -> YunlinkError | None:
        return self._poll_error

    def connect(self, ip: str, port: int) -> PeerConnection:
        return PeerConnection(call_native(self._core.connect, ip, port))

    async def connect_async(self, ip: str, port: int) -> PeerConnection:
        return await asyncio.to_thread(self.connect, ip, port)

    def open_session(self, peer: PeerConnection, node_name: str) -> Session:
        return Session(call_native(self._core.open_session, peer.id, node_name))

    async def open_session_async(self, peer: PeerConnection, node_name: str) -> Session:
        return await asyncio.to_thread(self.open_session, peer, node_name)

    def request_authority(
        self,
        peer: PeerConnection,
        session: Session,
        target: TargetSelector,
        source: ControlSource,
        lease_ttl_ms: int,
        allow_preempt: bool,
    ) -> None:
        call_native(
            self._core.request_authority,
            peer.id,
            session.session_id,
            target.to_native(),
            int(source),
            lease_ttl_ms,
            allow_preempt,
        )

    async def request_authority_async(
        self,
        peer: PeerConnection,
        session: Session,
        target: TargetSelector,
        source: ControlSource,
        lease_ttl_ms: int,
        allow_preempt: bool,
    ) -> None:
        await asyncio.to_thread(
            self.request_authority,
            peer,
            session,
            target,
            source,
            lease_ttl_ms,
            allow_preempt,
        )

    def release_authority(
        self, peer: PeerConnection, session: Session, target: TargetSelector
    ) -> None:
        call_native(
            self._core.release_authority, peer.id, session.session_id, target.to_native()
        )

    async def release_authority_async(
        self, peer: PeerConnection, session: Session, target: TargetSelector
    ) -> None:
        await asyncio.to_thread(self.release_authority, peer, session, target)

    def renew_authority(
        self,
        peer: PeerConnection,
        session: Session,
        target: TargetSelector,
        source: ControlSource,
        lease_ttl_ms: int,
    ) -> None:
        call_native(
            self._core.renew_authority,
            peer.id,
            session.session_id,
            target.to_native(),
            int(source),
            lease_ttl_ms,
        )

    async def renew_authority_async(
        self,
        peer: PeerConnection,
        session: Session,
        target: TargetSelector,
        source: ControlSource,
        lease_ttl_ms: int,
    ) -> None:
        await asyncio.to_thread(
            self.renew_authority, peer, session, target, source, lease_ttl_ms
        )

    def current_authority(self) -> AuthorityLease | None:
        lease = call_native(self._core.current_authority)
        if lease is None:
            return None
        return AuthorityLease(
            state=lease["state"],
            session_id=lease["session_id"],
            peer=PeerConnection(lease["peer_id"]),
        )

    def publish_goto(
        self,
        peer: PeerConnection,
        session: Session,
        target: TargetSelector,
        payload: GotoCommand,
    ) -> CommandHandle:
        result = call_native(
            self._core.publish_goto,
            peer.id,
            session.session_id,
            target.to_native(),
            payload.to_native(),
        )
        return CommandHandle(
            session_id=result["session_id"],
            message_id=result["message_id"],
            correlation_id=result["correlation_id"],
        )

    async def publish_goto_async(
        self,
        peer: PeerConnection,
        session: Session,
        target: TargetSelector,
        payload: GotoCommand,
    ) -> CommandHandle:
        return await asyncio.to_thread(self.publish_goto, peer, session, target, payload)

    def publish_vehicle_core_state(
        self,
        peer: PeerConnection,
        target: TargetSelector,
        payload: VehicleCoreState,
        session_id: int,
    ) -> None:
        call_native(
            self._core.publish_vehicle_core_state,
            peer.id,
            target.to_native(),
            payload.to_native(),
            session_id,
        )

    async def publish_vehicle_core_state_async(
        self,
        peer: PeerConnection,
        target: TargetSelector,
        payload: VehicleCoreState,
        session_id: int,
    ) -> None:
        await asyncio.to_thread(
            self.publish_vehicle_core_state, peer, target, payload, session_id
        )

    def _poll_loop(self) -> None:
        while not self._stop.is_set():
            try:
                configuration_payload = None
                poll_configuration = getattr(
                    self._core, "poll_configuration_response", None
                )
                if poll_configuration is not None:
                    configuration_payload = call_native(poll_configuration)
                event = call_native(self._core.poll_event)
            except YunlinkError as exc:
                self._poll_error = exc
                parsed = ErrorEvent(-1, str(exc))
                self._publish_event(parsed)
                break
            if configuration_payload is not None:
                configuration_response = coerce_configuration_response(
                    configuration_payload
                )
                if configuration_response is not None:
                    self._publish_configuration(configuration_response)
            if event is None and configuration_payload is None:
                time.sleep(0.01)
                continue

            if event is not None:
                parsed = coerce_event(event)
                if parsed is not None:
                    self._publish_event(parsed)

    def _publish_event(self, event: object) -> None:
        for subscriber in list(self._subscribers):
            subscriber.put(event)
        for loop, subscriber in list(self._async_subscribers):
            loop.call_soon_threadsafe(subscriber.put_nowait, event)

    def _publish_configuration(self, response: object) -> None:
        for subscriber in list(self._configuration_subscribers):
            subscriber.put(response)
        for loop, subscriber in list(self._async_configuration_subscribers):
            loop.call_soon_threadsafe(subscriber.put_nowait, response)
